sematic/resolvers/cloud_resolver.py

```diff
@@ -324,6 +333,57 @@ def _process_run_output(self, run_id: str) -> None:
 
         self._update_future_with_value(future, value)
 
+    def _resolution_did_fail(
+        self, error: Exception, reason: Optional[str] = None
+    ) -> None:
+        if not isinstance(error, ResolverRestartError):
+            super()._resolution_did_fail(error)
+            logger.error("Really early return!")
+            return
+
+        try:
+            new_root_future = self._root_future.calculator(
+                **self._root_future.kwargs
+            ).set(name=self._root_future.props.name, tags=self._root_future.props.tags)
+            runs, artifacts, edges = api_client.get_graph(
+                self._root_future.id, root=True
+            )
+            graph = Graph(
+                runs=runs,
+                artifacts=artifacts,
+                edges=edges,
+            )
+            rerun_from = graph.first_incomplete_run()
```

Comment:
This is pretty horrible. There does not exist the API that is needed to support this functionality cleanly:

Logically, rerun "from" is a way of invalidating a specific Run when rerunning a Resolution, so it is:
1. cloning completed Runs
2. invalidating the "from" Run, its downstream & upstream Runs
3. running every non-terminal Run

For this operation, you need steps 1. + 3., but the existing API is forcing you to do step 2. as well.

Please just modify the rerun API to not require a "from" Run (keep the restriction on user-facing API endpoints), and use that directly instead of copying the graph here just to search for an arbitrary Run, and invalidating it, even though it was not executed.

This would also avoid having to carry the git info around in the Resolver, as the existing rerun functionality already handles cloning the Resolution and scheduling a detached Resolver for you.

Reply:
I thought we had said this would be the behavior for now (cloning the entire run), as a quick solution.
If it IS possible – and not too hard – to do a rerun WITHOUT cloning resolved/ran runs, than it would indeed be preferable.

Reply:
To clarify: I'm not concerned if Resolved or Ran Runs are cloned or not. I want to avoid performing an extra copy of the Graph just to search for an incomplete Run to invalidate, just because the existing API requires passing a Run to invalidate.

Reply:
I don't think it's actually "horrible" for something we intend from the beginning to be a bandaid. Also, the overhead of cloning the graph is not that high; even user's obscenely large graphs are order of 100 nodes, which is nothing in terms of computation or memory required to clone.

Reply:
That said, I'll explore a "re-run without specifying an id to rerun from" option.

Reply:
implemented that option